File: src/test/java/cc/vileda/kalfor/RestApiMock.java
package cc.vileda.kalfor;

import io.vertx.core.json.JsonObject;
import io.vertx.rxjava.core.AbstractVerticle;
import io.vertx.rxjava.core.http.HttpServer;
import io.vertx.rxjava.ext.web.Router;


class RestApiMock extends AbstractVerticle
{
		private final int port;

		public RestApiMock(final int port)
		{
				this.port = port;
		}

		@Override
		public void start() throws Exception
		{
				final HttpServer httpServer = vertx.createHttpServer();
				final Router router = Router.router(vertx);
				router.route("/test").handler(routingContext ->
						routingContext.response()
								.putHeader("content-type", "application/json")
								.end(new JsonObject().put("foo", "bar").encodePrettily()));

				httpServer.requestHandler(router::accept).listen(port);
		}
}

File: library/src/test/kalfor/mock/RestApiMock.kt
package cc.vileda.kalfor.mock

import io.vertx.core.json.Json
import io.vertx.core.json.JsonObject
import io.vertx.core.logging.LoggerFactory
import io.vertx.rxjava.core.AbstractVerticle
import io.vertx.rxjava.ext.web.Router


internal class RestApiMock(private val port: Int) : AbstractVerticle() {
    override fun start() {
        val httpServer = vertx.createHttpServer()
        val router = Router.router(vertx)

        router.route("/test").handler { routingContext ->
            val headers = routingContext.request().headers()
            LOGGER.info(Json.encodePrettily(headers.names()))
            LOGGER.info(Json.encodePrettily(headers.names().map({ headers.getAll(it) })))
            routingContext.response().putHeader("content-type", "application/json").end(JsonObject().put("1foo", "1bar").encodePrettily())
        }

        router.route("/test2").handler { routingContext ->
            if (routingContext.request().getHeader("x-foo") == null) {
                routingContext.response().setStatusCode(500).end("missing header x-foo")
            } else {
                val headers = routingContext.request().headers()
                LOGGER.info(Json.encodePrettily(headers.names()))
                LOGGER.info(Json.encodePrettily(headers.names().map({ headers.getAll(it) })))
                routingContext.response().putHeader("content-type", "application/json").end(JsonObject().put("2foo", "2bar").encodePrettily())
            }
        }

        router.route("/testrdr").handler { routingContext ->
            routingContext.response()
                    .putHeader("Location", "/test")
                    .setStatusCode(301)
                    .end("Redirect!")
        }

        router.route("/testbroken").handler { routingContext ->
            routingContext.response()
                    .end("{Broken!")
        }

        httpServer.requestHandler({ router.accept(it) }).listen(port)
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(RestApiMock::class.java)
    }
}
